Add destination cluster info to response cookie

harishnelakurthi opened the pull request:
## Description
Add destination cluster a query is routed to in the cookie. More details here of the use-case here: https://github.com/trinodb/trino-gateway/issues/465

Fixes https://github.com/trinodb/trino-gateway/issues/465


## Release notes

```markdown
* The cookie data for requests to `/v1/statement` API would include the cluster to which a query got routed named as `trinoClusterHost`.
```

wendigo commented on gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
 
+    public void setIncludeClusterHostInResponse(boolean includeClusterHostInResponse)
+    {
+        this.includeClusterHostInResponse = includeClusterHostInResponse;
```
Config setters usually `return this` which allows for chaining

harishnelakurthi replied:
I see that all the other setters are returning void. Just wanted to be consistent. Let me know if we still want to change it.

wendigo commented on gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
+
+    public boolean getIncludeClusterHostInResponse()
+    {
+        return this.includeClusterHostInResponse;
```
drop `this.` - redundant

wendigo commented on gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
+        this.includeClusterHostInResponse = includeClusterHostInResponse;
+    }
+
+    public boolean getIncludeClusterHostInResponse()
```
should be: isIncludeClusterHostInResponse

wendigo commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -173,9 +177,14 @@ private void performRequest(
             future = future.transform(response -> recordBackendForQueryId(request, response), executor);
         }
 
+        Map<String, String> extraHeaders = new HashMap<>();
```
The mutable maps should be avoided.

This could be just:
`Map<String, String> extraHeaders = includeClusterInfoInResponse ? ImmutableMap.of(CLUSTER_HOST_HEADER, remoteUri.getHost()) : ImmutableMap.of()`

wendigo commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -206,10 +215,15 @@ private static String getRemoteTarget(URI remoteUri)
         Response.ResponseBuilder builder = Response.status(response.statusCode()).entity(response.body());
         response.headers().forEach((headerName, value) -> builder.header(headerName.toString(), value));
+
+        // Add any extra headers.
```
just: `extraHeaders.forEach(builder::header)`

willmostly commented:
@harishnelakurthi please sign the CLA so that we can take this forward, and address all comments. Please squash your commits as well.

harishnelakurthi commented:
@willmostly - I have signed the CLA and emailed a week ago. Also, I've addressed all the comments on the PR except there was an open question which I have responded to the reviewer. Let me know if anything is needed from my end.

willmostly commented:
@harishnelakurthi I think you may have missed my comment above. I would like to avoid adding a new header just to pass back information unrelated to the routing protocol. This should be a cookie. You can also update the logic such that the cookie is only added/updated in the response to a POST to `v1/statement`. Please update the PR title and description to reflect the change as well.

harishnelakurthi commented:
Thanks, @willmostly I have made the changes to pass this data via session cookie. Please review and let me know if any changes are needed. Also, updated the PR title and description. Will squash the commits before merge.

andythsu commented:
Can we change the PR name to "cluster" instead of "backend"

willmostly commented:
@harishnelakurthi you can use Airlift's codestyle in your IDE to avoid builds failing due to checkstyle https://github.com/airlift/codestyle

willmostly commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -169,13 +172,19 @@ private void performRequest(
         setupAsyncResponse(
                 asyncResponse,
-                future.transform(response -> buildResponse(response, oauth2GatewayCookie), executor)
+                future.transform(response -> buildResponse(response, oauth2GatewayCookie, finalSessionCookie), executor)
```
This is a bit more awkward than it needs to be because `getOAuth2GatewayCookie` returns an `ImmutableList`. Can you change it's return type to `List` and update the method to use some mutable implementation like `ArrayList`, then add your `trinoClusterHost` cookie to that list? Then you can revert the changes to `buildResponse`

willmostly commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -169,13 +172,19 @@ private void performRequest(
         if (statementPaths.stream().anyMatch(request.getUri().getPath()::startsWith) && request.getMethod().equals(HttpMethod.POST)) {
             future = future.transform(response -> recordBackendForQueryId(request, response), executor);
+            if (includeClusterInfoInResponse) {
+                // add trinoClusterHost to sessionCookies
```
not: comment here is unnecessary, the code is self-explanatory

willmostly reviewed (commented):
Looks pretty close % comments. 

Can you add a test for this functionality? `TestGatewayHaMultipleBackend` should have the infrastructure you need

willmostly commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -171,34 +174,39 @@ private void performRequest(
+            }
         }
 
+        ImmutableList<NewCookie> cookie = ImmutableList.copyOf(oauth2GatewayCookie);
```
alter `buildResponse` to accept a `List` instead of copying `oauth2GatewayCookie` to a new `ImmutableList`.

willmostly commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -160,7 +163,7 @@ private void performRequest(
         }
 
-        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        List<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
```
rename this to `cookies` since we will add a non-oauth2 cookie to the list

willmostly commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -145,6 +145,11 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Optional<Cookie> cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
+        assertThat(cookie).isNotEmpty();
```
We should test for a specific value to make sure the cluster host is set correctly. You can control which backend the request is sent to using the `X-Trino-Routing-Group` header. When you set 
```
.addHeader("X-Trino-Routing-Group", "scheduled")
```
then the backend should be `http://localhost:backend2Port`, when `X-Trino-Routing-Group` is unset you'll get `http://localhost:backend1Port`. You should also use the test to define the behavior when the initial client request already contains a cookie with the key `trinoClusterHost` - I expect the value should be overwritten. It will be cleaner to make a new test method.

willmostly reviewed (changes requested):
Minor code comments and some additional testing

willmostly commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -145,6 +145,7 @@ public void testQueryDeliveryToMultipleRoutingGroups()
                         .build();
         Response response1 = httpClient.newCall(request1).execute();
         assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+
```
nit: remove this additional newline

willmostly reviewed (approved):
I think this is ready. Please remove the stray newline from the `testQueryDeliveryToMultipleRoutingGroups` method and we can get this merged

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
         return routing;
     }
 
+    public void setIncludeClusterHostInResponse(boolean includeClusterHostInResponse)
```
Please move this getter and setter to the bottom for consistency with field order.

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/ha/config/HaGatewayConfiguration.java:
```diff
@@ -97,6 +98,16 @@ public RoutingConfiguration getRouting()
+        this.includeClusterHostInResponse = includeClusterHostInResponse;
+    }
+
+    public boolean isIncludeClusterHostInResponse()
```
Replace the order of getter and setter.

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -160,7 +162,7 @@ private void performRequest(
         }
 
-        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        List<NewCookie> cookies = getOAuth2GatewayCookie(remoteUri, servletRequest);
```
Please use `ImmutableList.Builder` instead.

harishnelakurthi replied:
This is intentionally a mutable list. Please check the code later which modifies the list for adding more cookies.

ebyhr replied:
I'm suggesting `ImmutableList.Builder` not `ImmutableList`

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -171,46 +173,50 @@ private void performRequest(
-    private ImmutableList<NewCookie> getOAuth2GatewayCookie(URI remoteUri, HttpServletRequest servletRequest)
+    private List<NewCookie> getOAuth2GatewayCookie(URI remoteUri, HttpServletRequest servletRequest)
     {
+        List<NewCookie> cookies = new ArrayList<>();
```
Please use `ImmutableList.Builder` instead.

harishnelakurthi replied:
Refer to my other comment.

ebyhr replied:
I'm suggesting `ImmutableList.Builder` not `ImmutableList`

ebyhr replied:
Actually, changes in this method aren't necessary. Please revert.

harishnelakurthi replied:
Please suggest alternative instead.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+            throws Exception
+    {
+        RequestBody requestBody =
+                RequestBody.create(MediaType.parse("application/json; charset=utf-8"), "SELECT 1");
```
Please avoid using deprecated `RequestBody.create` method. Also, pleas fix other waring in this test.

harishnelakurthi replied:
All the existing tests use the same. Can you please provide the right method to use here?

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        RequestBody requestBody =
+                RequestBody.create(MediaType.parse("application/json; charset=utf-8"), "SELECT 1");
+
+        // When X-Trino-Routing-Group is set in header, query should be routed to cluster under the
+        // routing group
```
nit: Unwrap.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        Response response1 = httpClient.newCall(request1).execute();
+        assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Optional<Cookie> cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
```
Please consider replacing `findAny()` with `collect(onlyElement())`.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Optional<Cookie> cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
+        assertThat(cookie).isNotEmpty();
+        String expectedHost = "localhost";
```
Inline this variable.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).findAny();
+        assertThat(cookie).isNotEmpty();
+        expectedHost = "localhost";
```
Inline this variable.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +158,49 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        // test with sending the request which includes trinoClusterHost in the cookie
+        // When X-Trino-Routing-Group is set in header, query should be routed to cluster under the
+        // routing group
+        Request request2 =
```
Please avoid number suffix as much as possible.

ebyhr commented:
Please squash commits into one.

ebyhr commented on gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java:
```diff
@@ -160,7 +162,9 @@ private void performRequest(
         }
 
-        ImmutableList<NewCookie> oauth2GatewayCookie = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        ImmutableList.Builder<NewCookie> cookieBuilder = new ImmutableList.Builder<>();
+        ImmutableList<NewCookie> oauthCookies = getOAuth2GatewayCookie(remoteUri, servletRequest);
+        oauthCookies.forEach(cookieBuilder::add);
```
```java
        ImmutableList.Builder<NewCookie> cookieBuilder = ImmutableList.builder();
        cookieBuilder.addAll(getOAuth2GatewayCookie(remoteUri, servletRequest));
```

ebyhr replied:
> ImmutableList.Builder<NewCookie> cookieBuilder = ImmutableList.builder();

This isn't addressed yet.

harishnelakurthi replied:
DOne

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +159,43 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        Response response1 = httpClient.newCall(request1).execute();
+        assertThat(response1.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(response1.request().url(), response1.headers());
+        Cookie cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(MoreCollectors.onlyElement());
```
Static import `onlyElement()`.

harishnelakurthi commented:
@ebyhr @willmostly Addressed all the comments.If there are no comments I can squash all the commits and would appreciate merging this in as it's been hanging around for a while.

ebyhr commented:
Please remove redundant commit body.

harishnelakurthi commented:
@ebyhr - I've addressed all the comments on this. Can this be merged please?

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+    public void testTrinoClusterHostCookie()
+            throws Exception
+    {
+        RequestBody requestBody = RequestBody.create("SELECT 1", MediaType.get("application/json; charset=utf-8"));
```
Please rebase and use `MEDIA_TYPE` constant.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .build();
+        Response responseOne = httpClient.newCall(requestWithoutCookie).execute();
```
Rename `responseOne` and `responseTwo` variables.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+        assertThat(responseOne.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(responseOne.request().url(), responseOne.headers());
+        Cookie cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(onlyElement());
+        assertThat(cookie.value()).isEqualTo("localhost");
+        // test with sending the request which includes trinoClusterHost in the cookie
```
nit: Add an empty line.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+                        .build();
+        Response responseTwo = httpClient.newCall(requestWithCookie).execute();
+        assertThat(responseTwo.body().string()).contains("http://localhost:" + routerPort);
+        cookies = Cookie.parseAll(responseTwo.request().url(), responseTwo.headers());
+        cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(onlyElement());
```
Please avoid reusing `cookies` variable.

ebyhr commented on gateway-ha/src/test/java/io/trino/gateway/ha/TestGatewayHaMultipleBackend.java:
```diff
@@ -158,6 +159,42 @@ public void testQueryDeliveryToMultipleRoutingGroups()
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .build();
+        Response responseOne = httpClient.newCall(requestWithoutCookie).execute();
+        assertThat(responseOne.body().string()).contains("http://localhost:" + routerPort);
+        List<Cookie> cookies = Cookie.parseAll(responseOne.request().url(), responseOne.headers());
+        Cookie cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(onlyElement());
+        assertThat(cookie.value()).isEqualTo("localhost");
+        // test with sending the request which includes trinoClusterHost in the cookie
+        // when X-Trino-Routing-Group is set in header, query should be routed to cluster under the routing group
+        Request requestWithCookie =
+                new Request.Builder()
+                        .url("http://localhost:" + routerPort + "/v1/statement")
+                        .addHeader("X-Trino-User", "test")
+                        .post(requestBody)
+                        .addHeader("X-Trino-Routing-Group", "scheduled")
+                        .addHeader("Cookie", "trinoClientHost=foo.example.com")
+                        .build();
+        Response responseTwo = httpClient.newCall(requestWithCookie).execute();
+        assertThat(responseTwo.body().string()).contains("http://localhost:" + routerPort);
+        cookies = Cookie.parseAll(responseTwo.request().url(), responseTwo.headers());
+        cookie = cookies.stream().filter(c -> c.name().equals("trinoClusterHost")).collect(onlyElement());
+        assertThat(cookie.value()).isEqualTo("localhost");
```
> When X-Trino-Routing-Group is set in header, query should be routed to cluster under the routing group
> ...
> test with sending the request which includes trinoClusterHost in the cookie when X-Trino-Routing-Group is set in header, query should be routed to cluster under the routing group

Values in assertions between `requestWithoutCookie` and `requestWithCookie` are the same. I don't understand what you want to express in this comment.

harishnelakurthi replied:
I'll fix the comment. The intent of this test is to make sure that the cookie is overriden correctly even though the request sends with some random value.

ebyhr reviewed (commented):
Looks good except for comments.

ebyhr commented:
@harishnelakurthi Please squash commits into one.

harishnelakurthi commented:
Thanks a lot @ebyhr for all the comments. I've squashed all the commits. Can you merge the PR if it's all good? Let me know if anything else needs to be done. Thanks!

ebyhr commented:
@harishnelakurthi Could you please use "rebase" instead of "merge"? 
![Screenshot 2024-09-27 at 6 55 52](https://github.com/user-attachments/assets/7f8b42cb-f67e-47d3-ab77-892fcf694d98)
